processing/src/main/java/org/apache/druid/segment/data/FixedIndexedWriter.java

```diff
@@ -197,13 +202,8 @@ private void readPage()
       {
         iteratorBuffer.clear();
         try {
-          if (numWritten - (pos - startPos) < PAGE_SIZE) {
-            int size = (numWritten - (pos - startPos)) * width;
-            iteratorBuffer.limit(size);
-            valuesOut.readFully((long) (pos - startPos) * width, iteratorBuffer);
-          } else {
-            valuesOut.readFully((long) (pos - startPos) * width, iteratorBuffer);
-          }
+          iteratorBuffer.limit(Math.min(PAGE_SIZE, (cardinality - pos) * width));
```

Comment:
cardinality includes the null value while numWritten did not (null values are not stored in the values, rather as a flag), shouldn't this be checking nulls? Additionally, it might make sense to adjust this computation to just precompute a `final int nullAdjust = hasNulls ?  1 : 0` so we can just do `cardinality - nullAdjust` in all the places that need to adjust

Reply:
`hasNext()` is implemented in terms of `totalCount` which had 1 added to it for the null case.  The old logic was achieving the handling of `null` stuff by doing `(pos - startPos)` which is following what you suggest.  For this line 205, `pos` has already been incremented once by the time it gets here in case that there are nulls (the first call to `next()` won't call `readPage()`).

Reply:
oops yea, i see that `startPos` was dropped :+1: